Add -lhHpw options to "zpool iostat" for avg latency, histogram, & queue stats

tonyhutter opened the pull request:
Update the zfs module to collect statistics on average latencies, queue sizes,
and keep an internal histogram of all IO latencies.  Along with this, update
"zpool iostat" with some new options to print out the stats:

**-l**: Include average IO latency stats:

```
 $ zpool iostat -vl
                         capacity     operations     bandwidth    total_wait     disk_wait    syncq_wait    asyncq_wait  scrub
pool                   alloc   free   read  write   read  write   read  write   read  write   read  write   read  write   wait
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool                  201G   171G      0  2.37K      0   137M      -   41ms      -    2ms      -   46ms      -    4ms      -
  A1                    100G  85.7G      0    249      0  3.89M      -    5ms      -    1ms      -    1us      -    4ms      -
  A2                    100G  85.6G      0    207      0  2.24M      -    5ms      -    1ms      -    1us      -    4ms      -
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A3                    174M   186G      0  1.92K      0   131M      -   49ms      -    2ms      -   47ms      -      -      -
cache                      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A4                   74.5K   186G      0     10      0  62.9K      -    2ms      -    1ms      -      -      -    1ms      -
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool2                 401G   343G      3  5.48K  4.99K  18.0M    1ms    1ms    1ms  413us   16us   25us      -    5ms      -
  raidz2                401G   343G      3  5.48K  4.99K  18.0M    1ms    1ms    1ms  413us   16us   25us      -    5ms      -
    A5                     -      -      1  1.37K  2.00K  4.49M    2ms    1ms    2ms  412us   26us   25us      -    5ms      -
    A6                     -      -      0  1.41K      0  4.47M      -    1ms      -  413us      -   25us      -    5ms      -
    A7                     -      -      0  1.28K      0  4.51M      -    1ms      -  460us      -   29us      -    5ms      -
    A8                     -      -      1  1.43K  3.00K  4.49M  196us    1ms  196us  370us    7us   23us      -    5ms      -
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
```

**-w**: Print out latency histograms:

```
$ zpool iostat -wp mypool

mypool        total_wait       disk_wait      sync_queue      async_queue 
latency       read   write    read   write    read   write    read   write   scrub
----------  ------  ------  ------  ------  ------  ------  ------  ------  ------
1ns              0       0       0       0       0       0       0       0       0
3ns              0       0       0       0       0       0       0       0       0
7ns              0       0       0       0       0       0       0       0       0
15ns             0       0       0       0       0       0       0       0       0
31ns             0       0       0       0       0       0       0       0       0
63ns             0       0       0       0       0       0       0       0       0
127ns            0       0       0       0       0       0       0       0       0
255ns            0       0       0       0       0       0       0       0       0
511ns            0       0       0       0       0       0       0       3       0
1us              0       0       0       0      28    1251    1276     321    4518
2us              0       0       0       0      91    4789    5939    1800    2697
4us              0       0       0       0      47    4246    1996    1113     158
8us              0       0       0       0      15     242     200      81      31
16us             0       0       0       0      21      66     119      47      25
32us             0       0       1       0      40      65      97      25      17
65us            72    1283     388   16099      33     150     175      21      37
131us          305     522    1755    5670       7     256     488      40      53
262us          706     569    5602    2431       4     556     830      88      66
524us         1000    1787    2717  741934       0    1057    1597     196     111
1ms           2525    3559  106792   40587       0    1941    3705     536     133
2ms           7845    5374   41675  152694       0    3304   10163     955     263
4ms          37600   16059   18739   48061       0    5919   30472    7669     179
8ms          55735   19017    1822   25880       0   11964   56032    5381      27
16ms         71730   24165      86    1543       0   16926   56598    6152       0
33ms          1980  100844      15     216       0   31901    1231   84587       0
67ms            97  811423       3      25       0   51720      76  741842       0
134ms            0   42166       0       1       0   39222       0     705       0
268ms            0    8191       0       0       0    7780       0      42       0
536ms            0      76       0       0       0       0       0      76       0
1s               0      29       0       0       0       0       0      29       0
2s               0      11       0       0       0       0       0      11       0
4s               0       7       0       0       0       0       0       7       0
8s               0      60       0       0       0       0       0      60       0
17s              0       0       0       0       0       0       0       0       0
34s              0       0       0       0       0       0       0       0       0
68s              0       0       0       0       0       0       0       0       0
137s             0       0       0       0       0       0       0       0       0
----------------------------------------------------------------------------------
```

**-h**: Help

**-H**: Scripted mode. Do not display headers, and separate fields by a single
    tab instead of arbitrary space.

**-q**: Include current number of entries in sync & async read/write queues,
    and scrub queue:

```
$ zpool iostat -vq
                         capacity     operations     bandwidth    syncq_read    syncq_write   asyncq_read  asyncq_write   scrubq_read
pool                   alloc   free   read  write   read  write   pend  activ   pend  activ   pend  activ   pend  activ   pend  activ
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool                  201G   171G  4.90K  1.44K   623M   164M      0      0      0      0     78     29      0      0      0      0
  A1                    100G  85.7G  2.42K      0   307M      0      0      0      0      0     78     29      0      0      0      0
  A2                    100G  85.6G  2.48K      0   316M      0      0      0      0      0      0      0      0      0      0      0
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A3                    791M   185G      0  1.44K      0   164M      0      0      0      0      0      0      0      0      0      0
cache                      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A4                     26K   186G      0      0      0      0      0      0      0      0      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
hutter2                 401G   343G  6.67K  3.61K   721M   329M      0      0    227    394      0     19      0      0      0      0
  raidz2                401G   343G  6.66K  3.61K   720M   329M      0      0    227    394      0     19      0      0      0      0
    A5                     -      -  1.70K    888   177M  79.1M      0      0    108     98      0     19      0      0      0      0
    A6                     -      -  1.64K    938   180M  84.1M      0      0     19     98      0      0      0      0      0      0
    A7                     -      -  1.71K    908   183M  81.6M      0      0     78     98      0      0      0      0      0      0
    A8                     -      -  1.62K    947   180M  84.1M      0      0     19     88      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
```
**-p**: Display numbers in parseable (exact) values.

Also, update iostat syntax to allow the user to specify specific vdevs
to show statistics for.  The three options for choosing pools/vdevs are:

Display a list of pools:
    _zpool iostat ... [pool ...]_

Display a list of vdevs from a specific pool:
    _zpool iostat ... [pool vdev ...]_

Display a list of vdevs from any pools:
    _zpool iostat ... [vdev ...]_

Lastly, allow zpool command "interval" value to be floating point.  This allows users to pass in fractions of a second to get faster polling intervals.  For example, to poll every half second:
    
`zpool iostat -v 0.5`

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+		 * Choose some sane default column sizes for printing the
+		 * raw numbers.
+		 */
+		[VS_DEFAULT] = 15, /* 1PB capacity */
```
1PB is too small

behlendorf replied:
I'm not so sure.  It's true that 1PB doesn't cover the worst case but it definitely covers the vast vast majority of pools.  The worst case here is that the headings don't line up perfectly but that probably isn't a big deal since this the expectation is that parsable data is being consumed by a script.

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+		 */
+		[VS_DEFAULT] = 15, /* 1PB capacity */
+		[VS_LATENCY] = 10, /* 1B ns = 10sec */
+		[VS_QUEUES] = 6,   /* 1M queue entries */
```
1M queue entries is also way too small

NB, the ZFS preference has been to use pretty-printed numbers (unfortunately all KiB) with a parseable option (-p) for getting the full integer. You'll never satisfy the 80-column crowd with this much information, so you might consider humans for 80-column consumers and programs for parseable.

behlendorf replied:
How deep have you seen these queues get?  My guess would have been that if any of them ever got 1M entries deep something was serious wrong.  :)

tonyhutter replied:
In my (very limited) testing on a high-end PC w/SSDs, I never saw a queue get above 10 entries when running "iozone -a".  That may not be representative though; I'm curious to see what kind of numbers other people get.

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -51,6 +51,7 @@
 #include <sys/fm/util.h>
 #include <sys/fm/protocol.h>
 #include <sys/zfs_ioctl.h>
+#include <math.h>
```
organizationally, you might consider moving all of this pretty-printing code to another file. zpool_main.c is already quite large and contains a lot of different logic. Thus it is becoming unwieldy in its size and scope.

tonyhutter replied:
Not sure I follow.  The niceprint low-level functions (zfs_nicenum/zfs_nicetime/__zfs_nicenum) are in libzfs_util.c, not zpool_main.c.  Are you suggesting moving them out of libzfs_util.c to a new file?

richardelling replied:
you are correct, I missed the file break. carry on

richardelling commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+	/* only toplevel vdevs have capacity stats */
+	if (vs->vs_space == 0) {
+		if (cb->cb_scripted)
+			printf("\t%c\t\%c", na, na);
```
"\%" appears to be a mistake

tonyhutter replied:
Good catch!  I'll fix it

richardelling commented on include/sys/fs/zfs.h:
```diff
@@ -741,6 +744,24 @@ typedef enum zpool_errata {
+#define	BIT(n)	(1ULL << (n))
+#endif
+
+enum vdev_stat_flags {VS_DEFAULT = 0, VS_LATENCY = 1, VS_QUEUES = 2,
```
style: this is not consistent with how enums are defined in the existing code

behlendorf replied:
The following would be consistent with the existing style and naming conventions.  You might also consider dropping the BIT() macro, it's not clear to me that is actually makes thing more readable.

```
typedef enum vdev_stat_flags {
        VS_DEFAULT_BIT  = 0,
        VS_LATENCY_BIT  = 1,
        VS_QUEUES_BIT   = 2,
        VS_HISTO_BIT    = 3,
        VS_FLAGS_COUNT  = 4,
} vdev_stat_flags_t;

/* Mask versions of all our flags */
#define VS_DEFAULT      BIT(VS_DEFAULT_BIT)
#define VS_LATENCY      BIT(VS_LATENCY_BIT)
#define VS_QUEUES       BIT(VS_QUEUES_BIT)
#define VS_HISTO        BIT(VS_HISTO_BIT)

#define VS_FLAGS_ALL (VS_LATENCY_BIT | VS_QUEUES_BIT | VS_HISTO_BIT | VS_FLAGS_COUNT);
```

tonyhutter replied:
Will fix

richardelling commented on lib/libzfs/libzfs_util.c:
```diff
@@ -596,27 +596,47 @@ zfs_strdup(libzfs_handle_t *hdl, const char *str)
  */
 void
-zfs_nicenum(uint64_t num, char *buf, size_t buflen)
+__zfs_nicenum(uint64_t num, char *buf, size_t buflen,
```
I'm not convinced this approach is better than simply leaving zfs_nicenum() as is, and adding zfs_nicetime() and zfs_nicenum10(). It seems like a lot of conditional code to try to share a single interface. Conditional code should be avoided when possible.

tonyhutter replied:
There's enough code reuse between nicenum and nicetime that I think the extra conditionals and lookup tables in __zfs_nicenum() are justified, although I agree it's borderline.  I'd argue that my incarnation of __zfs_nicenum does make it easier for future nicenum formats to be added though.  One day we might want to add a "ZFS_NICENUM_1000", for example.

behlendorf replied:
I have the sense that adding a `zfs_nicetime()` and `zfs_nicenum10()` would be the more maintainable thing to do here as well.  You might collapse provide a common `zfs_nicenum_impl()` function which takes the `k_units` as an argument and wrap it with `zfs_nicenum` and `zfs_nicenum10`.

tonyhutter replied:
I'm fine with renaming __zfs_nicenum() -> zfs_nicenum_impl() to follow convention.  After that you would have:

```
zfs_nicenum():  Wrapper function to niceprint base 1024
zfs_nicetime():  Wrapper function to niceprint time value
zfs_nicenum_impl(): Low-level function to niceprint values
```

However, I'm not clear on what `zfs_nicenum10` would do.  Is it for printing out the raw numerical values (i.e. the -p "parsable" option)?

richardelling replied:
sorry for the confusion, zfs_nicenum is base 2, so zfs_nicenum10 is my idea for base 10

richardelling replied:
I don't have a strong opinion of the head fake, but note that you've taken a 40 LOC function that fits nicely on one screen and grown it to 70 LOC, not counting the head fakes. Maybe just a second 40 LOC is easier to read?

NB, there are other places where zfs_nicenum is used to display things that should be base10, but that can be tackled as a separate PR

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2689,42 +2689,79 @@ vdev_accessible(vdev_t *vd, zio_t *zio)
+
+void
+vdev_get_stats(vdev_t *vd, vdev_stat_t *vs) {
+	spa_t *spa = vd->vdev_spa;
```
This is only used in the ASSERT below.  I'd suggest just replacing `spa` with `vd->vdev_spa` in the ASSERT.

tonyhutter replied:
Will fix

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2689,42 +2689,79 @@ vdev_accessible(vdev_t *vd, zio_t *zio)
+		 * We're a leaf.  Just copy our ZIO active queue stats in.  The
+		 * other leaf stats are updated in vdev_stat_update().
+		 */
+		for (t = 0; t < ZIO_PRIORITY_COUNT; t++)
```
Should be ZIO_PRIORITY_NUM_QUEUEABLE instead of ZIO_PRIORITY_COUNT to prevent  overflow.

tonyhutter replied:
Yea, I just built the code on Centos 7 and it caught the error.  Will fix, probably with an ARRAY_SIZE(vd->vdev_queue.vq_class).

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+
+	/* For each bit set in flags */
+	for (f = flags; f; f &= ~BIT(idx)) {
+		idx = ffsll(f) - 1;
```
Instead of `ffsll()` which is a glibc extension use `highbit64()`.  This isn't as efficient but it is more portable, in general we try to avoid gnu specific extensions.  Related to this we could definitely update our `highbit64()` implementation in user space, which is ZoL specific, to use `ffsll()` internally.

tonyhutter replied:
Sorry, forgot about highbit64().  Will fix

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+		/* Print our top labels centered over "read  write" label. */
+		for (i = 0; iostat_top_labels[idx][i] != NULL; i++) {
+			text_start = (int) ((rw_column_width)/2 -
+					strlen(iostat_top_labels[idx][i])/2);
```
[style] continued lines should be indented 4 spaces.

tonyhutter replied:
Will fix

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+
+
+static void
+__print_iostat_separator(iostat_cbdata_t *cb, unsigned int force_column_width)
```
[style] the existing convention for wrapping a function like this is to append `_impl` to the function name.  So in this case it would be `print_iostat_separator_impl()`.  I made this same mistake early on as well which is why you still encounter the __ prefix for this in some places.

That said, I doesn't look like we need to provide a wrapper here.  These functions are all private to `zpool_main.c` so we can update all the callers and don't need to maintain a legacy interface.

tonyhutter replied:
Will remove wrapper function, and convert all __ functions to _impl or _cb.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+		for (i = 0; i < ARRAY_SIZE(calcvs->vs_queue_histo[0]); i++)
+			calcvs->vs_queue_histo[j][i] = scale *
+			(newvs->vs_queue_histo[j][i] -
+			oldvs->vs_queue_histo[j][i]);
```
[style] The existing style is a 4 space indent to continue a line (not a tab).  You'll want to go back and apply this consistently throughout the patch.

tonyhutter replied:
Will fix all instances

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+		max = MAX(max, array[i]);
+
+	return (max);
+}
```
It may make sense to move generic functions which don't take a `iostat_cbdata_t` as an argument in to `cmd/zpool/zpool_util.c`.

tonyhutter replied:
Will do

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+	if (!calcvs) {
+		printf("Not enough memory to print\n");
+		return (ret);
+	}
```
You can safely drop the `!calcvs` check here.  `safe_malloc()` performs it internally and will log and error and exit unconditionally if the allocation fails.

tonyhutter replied:
Nice!  Will fix

behlendorf commented on include/sys/fs/zfs.h:
```diff
@@ -741,6 +744,24 @@ typedef enum zpool_errata {
+#define	VS_DEFAULT_M	BIT(VS_DEFAULT)
+#define	VS_LATENCY_M	BIT(VS_LATENCY)
+#define	VS_QUEUES_M	BIT(VS_QUEUES)
+#define	VS_HISTO_M	BIT(VS_HISTO)
```
These names may be a bit to generic.  It's conceivable that we'll want to add additional statistics, for example a per-vdev histogram which measures something else we're interested in.

tonyhutter replied:
Good point, I can see us adding a request size histogram at some point.  I want to be careful though, since the macros are used a lot, and the longer names will push code towards the 80-char limit.  At the least, I'll rename VS_HISTO to something more specific, since we'll probably of adding more histos.

behlendorf replied:
A request size histogram would be sweet.

richardelling replied:
Unfortunately, due to coalescing (see vdev_queue_aggregate()), vdev-level performance is less useful than disk or vfs level. It becomes very difficult to correlate a vfs-level request to one or more disk I/Os, and visibility at the vdev-level doesn't help much. That said, I do like the approach and think that we can reuse this work at other levels. Thanks

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+static uint64_t
+get_stat_flags(zpool_list_t *list)
+{
+	uint64_t mask = -1; /* Start off with all 0xFFs */
```
It would be better to initialize this to `VS_CURRENT_FLAGS`.  This way the intersection of the bits provided by the utilities and the bits returned by the kernel modules describe the mutually supported flags.  This way you can never end up with a bit set which the utilities don't understand because of what the module might have returned.

tonyhutter replied:
Good idea, will fix

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+ * only the flags that are supported on all pools remain.
+ */
+static int
+__get_stat_flags(zpool_handle_t *zhp, void *data)
```
Often in the existing code functions which are used as callback use a `_cb` suffix.  So in this case I'd suggest calling this `get_stat_flags_cb()`.

tonyhutter replied:
Will fix

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+	 * __get_stat_flags() will lop off bits from "mask" until only the
+	 * flags that are supported on all pools remain.
+	 */
+	pool_list_iter(list, B_FALSE, __get_stat_flags, &mask);
```
It's not clear to me why the existing `zpool_iter()` function couldn't be used here.

tonyhutter replied:
Nice, I didn't know about `zpool_iter()`. I'll try that

tonyhutter replied:
I see that `zpool_iter()` takes in a `libzfs_handle_t`, while `pool_list_iter()` takes a` zpool_list_t`.  Since I only have a `zpool_list_t` in `zpool_do_iostat()`, I'm going to stick with `pool_list_iter()`.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+}
+
+/*
+ * This is the equivalent of for_all_pools() for vdevs.  It iterates thorough
```
You mean `for_each_pool()`, and this iterate logic should be part of `cmd/zpool/zpool_iter.c`

tonyhutter replied:
Good catch, will fix and will move to `zpool_iter.c`

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+ * return 0.  If pool_name is NULL, then search all pools.
+ */
+static int
+are_vdevs_in_pool(int argc, char **argv, char *pool_name,
```
It may make sense to move some of this vdev handling code over to `cmd/zpool/zpool_vdev.c` with the rest of it.

tonyhutter replied:
Unfortunately in this case it's specific to `zpool_main.c`, since `__is_vdev()` expects "cb_data" to be an iostat_cbdata_t.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2836,12 +3393,263 @@ get_timestamp_arg(char c)
+
+	/*
+	 * __get_stat_flags() will lop off bits from "mask" until only the
+	 * flags that are supported on all pools remain.
```
Currently all pools share the same kernel modules so it's not clear to me exactly how different pools could support different flags.  However, I'm all for making this as generic as possible because I have no doubt that we might eventually find a case for that, so I'm good with this.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2915,20 +3811,52 @@ zpool_do_iostat(int argc, char **argv)
+		cb.cb_flags = (histo << VS_HISTO);
+	} else {
+		cb.cb_flags = (1 << VS_DEFAULT) | (latency << VS_LATENCY) |
+				(queues << VS_QUEUES);
```
I think making the `VS_LATENCY` -> `VS_LATENCY_BIT` change would make this code clearer.  You'd then be able to clearly OR in or out the desired bits.

tonyhutter replied:
The VS_LATENCY, VS_QUEUES and VS_HISTO macros have existing bitmask equivalents macros defined as VS_LATENCY_M, VS_QUEUES_M, and VS_HISTO_M.  I think I can use them to make this code easier to read.

behlendorf commented on man/man8/zpool.8:
```diff
@@ -95,7 +95,8 @@ zpool \- configures ZFS storage pools
 .LP
 .nf
-\fBzpool iostat\fR [\fB-T\fR d | u ] [\fB-gLPvy\fR] [\fIpool\fR] ... [\fIinterval\fR[\fIcount\fR]]
+\fB\fBzpool iostat\fR [\fB-T\fR \fBd\fR | \fBu\fR] [\fB-ghHLpPvy\fR] [\fB-G\fR|[\fB-lq\fR]] [[\fIpool\fR ...]|[\fIpool vdev\fR ...]|[\fIvdev\fR ...]] [\fIinterval\fR[\fIcount\fR]]\fR 
```
This needs an explicit line break so it formats correctly at 80 columns.

tonyhutter replied:
Will fix

behlendorf commented on man/man8/zpool.8:
```diff
@@ -1664,11 +1665,22 @@ Scan using the default search path, the libblkid cache will not be consulted.  A
+interval, the statistics are printed every \fIinterval\fR seconds until
+\fBCtrl-C\fR is pressed. If \fIcount\fR is specified, the command exits after
+\fIcount\fR reports are printed.  The first report printed is always the
+statistics since boot, irregardless of whether \fIinterval\fR and \fIcount\fR
```
s/, irregardless/ regardless/

tonyhutter replied:
Will fix

behlendorf commented on man/man8/zpool.8:
```diff
@@ -1697,6 +1709,28 @@ Display vdev GUIDs instead of the normal device names. These GUIDs can be used i
+.RS 12n
+.rt
+Display help (usage)
+.RE
```
I'd consider dropping this from the man page to be consistent with all the other sub-commands.  It's OK that it work but I think it's a little distracting to add here with all the other iostat specific commands.

tonyhutter replied:
I'll remove it from the man page

behlendorf commented on man/man8/zpool.8:
```diff
@@ -1736,13 +1781,146 @@ Verbose statistics. Reports usage statistics for individual \fIvdevs\fR within t
+.ad
+.RS 20n
+.rt
+Amount of time IO spent in synchronous priority queues.
```
Does this include disk time?  My reading of this is that it doesn't.

tonyhutter replied:
Correct, sync_queue does not include disk time, only time in queue.  I'll add a line to make it clearer.

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2689,42 +2689,79 @@ vdev_accessible(vdev_t *vd, zio_t *zio)
+	}
+
+	ASSERT(spa_config_held(spa, SCL_ALL, RW_READER) != 0);
+	mutex_enter(&vd->vdev_stat_lock);
```
The `vdev_stat_lock` should be taken before the `bcopy()` above.

tonyhutter replied:
Good catch, will fix

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2689,42 +2689,79 @@ vdev_accessible(vdev_t *vd, zio_t *zio)
  */
 void
-vdev_get_stats(vdev_t *vd, vdev_stat_t *vs)
+__vdev_get_stats(vdev_t *vd, vdev_stat_t *vs)
```
s/__vdev_get_stats/vdev_get_stats_impl()/

Am I correct in understanding this version must now be fully recursive because the collection of the stats is now done in the leaves whereas before it was at multiple levels of the tree?

tonyhutter replied:
"__" -> "impl": will fix.

Correct, it's done in the leaves.  The major reason for the change was to make it easy to aggregate histogram numbers to the top level vdevs/pools.  IIRC, before my patch the mid level vdevs, like raidz, wouldn't aggregate the histo numbers from their children correctly, since it was only aggregating from the direct children underneath a root vdev.

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2811,8 +2848,19 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
+		 * The bytes/ops/histograms are recorded at the leaf level and
+		 * aggregated into the higher level vdevs in vdev_get_stats().
+		 */
+		if (vd->vdev_ops->vdev_op_leaf) {
```
This bit is what restricts the updates to the leaves.

behlendorf commented on module/zfs/vdev_disk.c:
```diff
@@ -570,7 +570,7 @@ __vdev_disk_physio(struct block_device *bdev, zio_t *zio, caddr_t kbuf_ptr,
 	vdev_disk_dio_get(dr);
 	if (zio)
-		zio->io_delay = jiffies_64;
+		zio->io_disk_delay = gethrtime();
```
We should pull this timing logic out of `vdev_disk.c` and up a level.  This has at least two major benefits.

1) It will then also work properly for file vdevs
2) It can be updated to include the `vdev_queue_aggregate()` time which in the current implementation isn't account for in the disk time.  That's vdev_queue_io_to_issue() to vdev_queue_io_done().

tonyhutter replied:
`io_disk_delay` is only use for a single error message in `vdev_disk.c`:

  ```
struct zio {
...
        hrtime_t        io_delay;       /* Device access time (disk or */
                                        /* file). */
        hrtime_t        io_disk_delay;  /* Same as io_delay, but for disks */
                                        /* only.  Used for printing the */
                                        /* delay in error messages in */
                                        /* vdev_disk.c */
```
`io_delay` is what you're looking for, and it is calculated at a higher level (in `zio_vdev_io_start()`/`zio_vdev_io_done()`).  At that level, it does include file-based vdevs.

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2811,8 +2848,19 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
+			vs->vs_bytes[type] += psize;
+			vs->vs_queue_histo[zio->io_priority]
+						[HISTO(zio->io_delta)]++;
+			vs->vs_disk_histo[type][HISTO(zio->io_delay)]++;
```
Shouldn't this be io_disk_delay?

tonyhutter replied:
io_delay is correct (see previous comment)

behlendorf commented on module/zfs/vdev.c:
```diff
@@ -2811,8 +2848,19 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
+			vs->vs_ops[type]++;
+			vs->vs_bytes[type] += psize;
+			vs->vs_queue_histo[zio->io_priority]
+						[HISTO(zio->io_delta)]++;
```
The io_delta time I believe includes the disk and queue time.  It's updated in `vdev_queue_io_done()`.

tonyhutter replied:
I was going off the comment in the struct, which I (perhaps incorrectly) interpreted as queue time only:
```
struct zio {
    ...
    hrtime_t        io_delta;       /* vdev queue service delta */
```
However, looking further at the code it doesn't seem right.  I would expect the queue ending time to be calculated at the very beginning of `zio_vdev_io_start()`, since that would include the time spent in the active queue.  I will try using that calculation instead.

behlendorf commented:
OK, I'm done!  Things look really nice and I personally like how this is shaping up!  Just needs to a few rounds of review and refinement so we can get it right.

sempervictus commented:
Not sure how important this is, but it looks like this PR generates a compiler warning:
```
  CC       util.lo
util.c: In function 'show_vdev_stats':
util.c:135:1: warning: the frame size of 6160 bytes is larger than 1024 bytes [-Wframe-larger-than=]
```

behlendorf commented:
@sempervictus should be fixed when this patch is next rebased and updated.

tonyhutter commented:
Thanks everyone for all the comments.  I've already got the build errors fixed in my local branch, and am working on implementing the other fixes you've suggested.

tonyhutter commented:
This latest update fixes the things I said I would fix, plus a few other problems I caught.

- Fix build issues under Centos 7 (array overflow / stack frame overflow)
- Fix 80 char line wrap on man page
- Fix unneeded '\' print in print_iostat_default()
- Rename underscore-underscore functions to _impl or _cb
- Use 4 spaces for continuation lines
- Remove BIT() macro
- Add lowbit64() in place of ffsll()
- Don't interpret GUIDs as time interval values
- Don't misinterpret pools/vdevs starting with a number as an interval value
- Print correct values for queue histo async reads
- Rename VS_HISTO to more specific VS_L_HISTO (latency histo)
- Move array64_max() and for_each_vdev() to library files
- Remove unneeded NULL check after safe_malloc()
- Use correct locking in vdev_get_stats()
- Use correct calculation for queing time.  Before, I was erroneously using zio->io_delta, which included both the queuing and disk time.
- Fix VS_* enums to use correct style
- Cleaned up an ASSERT on vd->vdev_spa
- set mask = VS_CURRENT_FLAGS in get_stat_flags()
- Make the code to setup cb->cb_flag clearer
- Remove "-h" help section from man page
- Update sync_queue section of man page to be clearer about disk time
- Add -lm to zpool to fix linker error
- Decouple ZPOOL_CONFIG_VDEV_STATS_FLAGS flags from zpool iostat flags. The flags in ZPOOL_CONFIG_VDEV_STATS_FLAGS have an almost 1:1 relationship with the zpool iostat display flags.  I say almost, because the VS_L_HISTO flag in the module corresponds to two displayed stats in zpool iostat: this histo itself, and the average latencies calculated from it.  In order to handle this the right way, we need to decouple the module's stats from zpool iostat stats.

@behlendorf - We talked offline about updating "zdb -s" with some of the new stats, but it turned out to be more complicated than just adding a few extra nicenum() prints in show_vdev_stats().  The problem is that the average latency isn't stored in the vdev_stats_t, but rather calculated from the histogram values.  We can't do a "nicenum(vs->vs_average_latency, ..)", for example.  We can easily niceprint the queues, but without the latency values it would just look out of place, so I ended up not adding it.

tonyhutter commented:
Wait on pulling this - I want to investigate if nvlists will work better for passing back the stats instead of expanding vdev_stats_t.

tonyhutter commented:
This is the latest incarnation of the stats code.  The biggest change is using nvlists for each individual histogram, rather than expanding vdev_stat_t.  This removes the need for ZPOOL_CONFIG_VDEV_STATS_FLAGS, and also makes the design more flexible.  We can grow/shrink the number of buckets without having to change the "zpool" command, for example. It also makes it easier to add stats in the future.  Lastly, I fixed a corner case where niceprint would show "1000us" instead of "1ms".

behlendorf commented on cmd/zpool/zpool_iter.c:
```diff
@@ -250,3 +250,70 @@ for_each_pool(int argc, char **argv, boolean_t unavail,
+		if (nvlist_lookup_nvlist_array(nv, list[i], &child,
+		    &children) == 0) {
+			for (c = 0; c < children; c++) {
+				uint64_t ishole = B_FALSE;
```
Assignment of a `boolean_t` to a `uint64_t`.  Best to just set this to 0 rather than mix types.

behlendorf commented on cmd/zpool/zpool_iter.c:
```diff
@@ -250,3 +250,70 @@ for_each_pool(int argc, char **argv, boolean_t unavail,
+	}
+
+	if (nvlist_lookup_string(nv, ZPOOL_CONFIG_TYPE, &type) != 0)
+		goto end;
```
This `goto end` doesn't appear to buy us anything over a normal `return (ret);`.  We have no additional state to unwind in this function so I'd opt for the return.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -237,8 +255,9 @@ get_usage(zpool_help_t idx) {
-		return (gettext("\tiostat [-gLPvy] [-T d|u] [pool] ... "
-		    "[interval [count]]\n"));
+		return (gettext("\tiostat [-ghHLpPvy] [[-lq]|-G] [-s name] "
+		    "[-T d|u] [-y]"
```
The `[-y]` here is redundant and you're missing a `\n` before the `[pool...` chunk.

behlendorf commented on man/man8/zpool.8:
```diff
@@ -1736,13 +1771,149 @@ Verbose statistics. Reports usage statistics for individual \fIvdevs\fR within t
+
+.RE
+
+All histogram buckets are power-of-two sized.  The times labels are the end
```
s/times labels/time labels/

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2479,61 +2498,595 @@ zpool_do_import(int argc, char **argv)
+    unsigned int column_size, boolean_t scripted)
 {
 	char buf[64];
+	zfs_nicenum_impl(value, buf, sizeof (buf), format);
```
While not a hard and fast rule, `*_impl` functions are typically declared static and not exposed with a public interface.  That way we can safely modify their prototype if we need to in the future without breaking anything.  So if we really want to able to pass a format specifier let's add a new tiny `zfs_nicenum_format()` wrapper and use that.

tonyhutter replied:
Or even simpler, rename zfs_nicenum_imp() ->  zfs_nicenum_format()

behlendorf replied:
That'd work too, and we really wouldn't be locking ourselves in to anything since once we add `zfs_nicenum_format()` we'll need to maintain that interface.

behlendorf commented on include/libzfs.h:
```diff
@@ -749,10 +749,18 @@ extern int zfs_unshareall(zfs_handle_t *);
  */
 extern void zfs_nicenum(uint64_t, char *, size_t);
+extern void zfs_nicenum_impl(uint64_t num, char *buf, size_t buflen,
+    enum zfs_nicenum_format type);
```
Avoid internal `*_impl()` functions in public headers.  Mentioned above.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -596,27 +596,49 @@ zfs_strdup(libzfs_handle_t *hdl, const char *str)
 void
-zfs_nicenum(uint64_t num, char *buf, size_t buflen)
+zfs_nicenum_impl(uint64_t num, char *buf, size_t buflen,
+					enum zfs_nicenum_format format)
```
Extra whitespace here.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2479,61 +2498,595 @@ zpool_do_import(int argc, char **argv)
+calc_default_iostats(vdev_stat_t *oldvs, vdev_stat_t *newvs,
+    vdev_stat_t *calcvs)
+{
+	int i;
```
[style] Not a rule which is always followed by usually an empty line is left between variable declarations and the body of a block.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2479,61 +2498,595 @@ zpool_do_import(int argc, char **argv)
+	for (j = 0; j < len; j++) {
+		rc = nvlist_lookup_uint64_array(newnv, names[j],
+		    (uint64_t **)&newnva[j].data, &newnva[j].len);
+		verify(rc == 0);
```
Often a `VERIFY0(nvlist_lookup_uint64_array...);` is used for this allowing you to remove the `rc` variable entirely.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2479,61 +2498,595 @@ zpool_do_import(int argc, char **argv)
+		print_solid_separator(entire_width);
+}
+
+/* Calculate the average latency of a power-of-two latency histogram */
```
[style] For function prototype block comments use the multi-line format even if it's just a single line.  This doesn't apply to comments inside functions.

```
/*
 * Calcualte...
 */
```

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2844,12 +3441,262 @@ get_timestamp_arg(char c)
+static uint64_t
+get_stat_flags(zpool_list_t *list)
+{
+	uint64_t mask = -1;
```
I think I commented on this one before, but `mask` should be initialized to only those bits that the utilities support.  That way older utilities with newer kernel modules we never see flags they can't handle.  You'll want to update the comment above `get_stat_flags_cb()` as well.

tonyhutter replied:
`get_stat_flags()` will return a bitmask of the IOS_* bits which are supported by both the client and module (see `get_stat_flags_cb()`).  I need to update the function comments to make that clear though; right now it still has the old comment from when `get_stats_flags()` was strictly returning the bits supported by the module only.

In interim versions of the code I did have `mask = VS_CURRENT_FLAGS`, but later removed it when I switched to using per-nvlist histograms (which removed the VS_CURRENT_FLAGS macro since it was no longer needed).  VS_CURRENT_FLAGS was basically the IOS_ALLFLAGS_M mask you described.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2844,12 +3441,262 @@ get_timestamp_arg(char c)
+	 * of all the nvlists a flag requires.
+	 */
+	static const char *type_to_nvlist[IOS_COUNT][10] = {
+		[IOS_DEFAULT] = {ZPOOL_CONFIG_VDEV_STATS, NULL},
```
I'd suggest putting each ZPOOL_CONFIG_* entry and the NULL on its own line to improve readability.  It helps eliminate any confusion when looking at this code.

```
                [IOS_DEFAULT] = {
                    ZPOOL_CONFIG_VDEV_STATS, 
                    NULL },
```

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2844,12 +3441,262 @@ get_timestamp_arg(char c)
+			continue;
+
+		/* Start off by assuming the flag is supported, then check */
+		flags |= (1ULL << j);
```
This could also be accomplished by initializing flags to the known set of flags supported by the utility.  Say an `IOS_ALLFLAGS_M` or some such.

tonyhutter replied:
I'd like to avoid defining a IOS_ALLFLAGS_M mask if possible.  While it would make the line more readable, it would also be something we would have to update with every new stat added (and I do plan on adding more stats in the not-to-distant future...)

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2844,12 +3441,262 @@ get_timestamp_arg(char c)
+	};
+
+	config = zpool_get_config(zhp, NULL);
+	verify(nvlist_lookup_nvlist(config, ZPOOL_CONFIG_VDEV_TREE,
```
I agree, moving to an nvlist for this will leave us a lot more flexibility in the long term.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -629,13 +651,51 @@ zfs_nicenum(uint64_t num, char *buf, size_t buflen)
+	zfs_nicenum_impl(num, buf, buflen, ZFS_NICENUM_TIME);
+}
+
+
```
Adding a `zfs_niceraw()` for completeness might be useful.

behlendorf commented on lib/libzpool/kernel.c:
```diff
@@ -1071,6 +1071,50 @@ highbit64(uint64_t i)
+/*
+ * Find lowest one bit set.
+ * Returns bit number + 1 of lowest bit that is set, otherwise returns 0.
+ * This is basically a reimplementation of ffsll(), which is GNU specific.
```
You still have a few `ffsll()` callers which should be updated to use `lowbit64()`.

behlendorf commented on lib/libzpool/util.c:
```diff
@@ -76,6 +76,12 @@ show_vdev_stats(const char *desc, const char *ctype, nvlist_t *nv, int indent)
+	if (!v0) {
+		printf("%s: not enough memory\n", __func__);
+		return;
+	}
```
`umem_zalloc(sizeof (*v0), UMEM_NOFAIL);` is an option here since this is the user bit of libzpool.

behlendorf commented on module/zfs/spa.c:
```diff
@@ -3210,6 +3210,62 @@ spa_add_spares(spa_t *spa, nvlist_t *config)
+		    vsx_queue_histo[ZIO_PRIORITY_SCRUB])},
+	};
+
+	vsx = kmem_alloc(sizeof (*vsx), KM_SLEEP);
```
This is a Solaris'ism but kmem_alloc(.., KM_SLEEP) is guaranteed to never fail.  Period.  We'll either get the memory or block here forever, so you can drop the error handling.

behlendorf commented on include/sys/fs/zfs.h:
```diff
@@ -528,6 +529,22 @@ typedef struct zpool_rewind_policy {
 #define	ZPOOL_CONFIG_DTL		"DTL"
 #define	ZPOOL_CONFIG_SCAN_STATS		"scan_stats"	/* not stored on disk */
 #define	ZPOOL_CONFIG_VDEV_STATS		"vdev_stats"	/* not stored on disk */
+#define	ZPOOL_CONFIG_VDEV_STATS_FLAGS	"vdev_stats_flags"	/* not stored */
```
This is unused.

behlendorf commented on module/zfs/vdev_label.c:
```diff
@@ -208,14 +208,77 @@ vdev_label_write(zio_t *zio, vdev_t *vd, int l, void *buf, uint64_t offset,
+	vdev_stat_ex_t *vsx;
+
+	vs = kmem_alloc(sizeof (*vs), KM_SLEEP);
+	ASSERT(vs);
```
See above, this cannot fail.

behlendorf commented on module/zfs/vdev_label.c:
```diff
@@ -208,14 +208,77 @@ vdev_label_write(zio_t *zio, vdev_t *vd, int l, void *buf, uint64_t offset,
+
+	vdev_get_stats_ex(vd, vs, vsx);
+	fnvlist_add_uint64_array(nv, ZPOOL_CONFIG_VDEV_STATS,
+			(uint64_t *)vs, sizeof (*vs) / sizeof (uint64_t));
```
[style] whitespace

behlendorf commented on module/zfs/vdev_label.c:
```diff
@@ -208,14 +208,77 @@ vdev_label_write(zio_t *zio, vdev_t *vd, int l, void *buf, uint64_t offset,
+
+	/* Add extended stats nvlists */
+	fnvlist_add_uint64_array(nv, ZPOOL_CONFIG_QUEUE_ACTIVE,
+	    vsx->vsx_queue_active, ARRAY_SIZE(vsx->vsx_queue_active));
```
It's not inconceivable additional queues might be added some day (or maybe removed?).  Using a uint64_t nvlist entry per-queue type would allow us to future proof this.

Have you given any thought to placing all of these new extended stats in to a new `ZPOOL_CONFIG_VDEV_STATS_EX` which is itself a nvlist.  You can add nvlists to nvlists with `fnvlist_add_nvlist()`.  This could make it easier to pass the extended stats around and allow them to be easily added/removed without serious user/kernel compatibility issues.

tonyhutter replied:
I think the per-queue nvlists are a good idea.  I'll update the code for that.

I wasn't aware of the nvlist of nvlists.  Sounds like a nice way to group the stats.  Let me look into how that will work with the existing code before I commit to it.

behlendorf commented on include/sys/fs/zfs.h:
```diff
@@ -528,6 +529,22 @@ typedef struct zpool_rewind_policy {
+#define	ZPOOL_CONFIG_SYNC_W_LAT_HISTO	"sync_w_lat_histo"
+#define	ZPOOL_CONFIG_ASYNC_R_LAT_HISTO	"async_r_lat_histo"
+#define	ZPOOL_CONFIG_ASYNC_W_LAT_HISTO	"async_w_lat_histo"
+#define	ZPOOL_CONFIG_SCRUB_LAT_HISTO	"scrub_histo"
```
Perhaps the should be prefixed `ZPOOL_CONFIG_VDEV_*` for clarity.  What do you think.

tonyhutter replied:
Sounds good

behlendorf commented on module/zfs/spa.c:
```diff
@@ -3261,6 +3317,7 @@ spa_add_l2cache(spa_t *spa, nvlist_t *config)
 			    ZPOOL_CONFIG_VDEV_STATS, (uint64_t **)&vs, &vsc)
 			    == 0);
 			vdev_get_stats(vd, vs);
+			spa_update_l2cache_stats_ex(vd, l2cache[i]);
```
This feels really clunky, surely we can use `vdev_config_generate_stats()` for this.  Right?

tonyhutter replied:
Maybe, but I'm not sure.  If you look at the case of the regular stats, `spa_add_l2cache()` looks up the vdev_stats struct from the existing nvlist, and updates it with new data, rather than adding a new vdev_stats nvlist (which is what `vdev_config_generate_stats()` does).  I assume the code did it this way for a reason, so I did it the same way for the extended stats.  I will have to test and see if `vdev_config_generate_stats()` will work here instead.

behlendorf commented on include/sys/zio_priority.h:
```diff
@@ -29,8 +29,9 @@ typedef enum zio_priority {
-
-	ZIO_PRIORITY_NOW		/* non-queued i/os (e.g. free) */
+	ZIO_PRIORITY_NOW,		/* non-queued i/os (e.g. free) */
+	ZIO_PRIORITY_COUNT,		/* Count of all priorities. Leave */
```
There's no need to add this, the patch should be using `ZIO_PRIORITY_NUM_QUEUEABLE` throughout (even if it is a terrible name).  There exactly this many `vdev_queue_class_t`'s defined in a `vdev_queue_t`.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -2704,6 +2707,7 @@ zio_vdev_io_start(zio_t *zio)
 		/*
 		 * The mirror_ops handle multiple DVAs in a single BP.
 		 */
+		zio->io_delay = gethrtime();
```
I think both this line and the`zio->io_delay = gethrtime()` should be dropped.  In the first case it's redundant and in the mirror case the mirror-type vdev will submit new zios to the leaves.  We just want to count physical zio's at the leaves to avoid double counting.  But please double check this.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -2810,11 +2815,14 @@ zio_vdev_io_done(zio_t *zio)
 	vdev_ops_t *ops = vd ? vd->vdev_ops : &vdev_mirror_ops;
 	boolean_t unexpected_error = B_FALSE;
 
+
```
[style] extra line

behlendorf commented on include/sys/zio.h:
```diff
@@ -421,7 +421,12 @@ struct zio {
+	hrtime_t	io_disk_delay;	/* Same as io_delay, but for disks */
+					/* only.  Used for printing the */
+					/* delay in error messages in */
+					/* vdev_disk.c */
```
Let's get rid of `io_disk_delay` entirely, with your updated infrastructure it doesn't retain any real value and will only serve as a future point of confusion.

tonyhutter commented:
Thanks for the feedback @behlendorf.  Any comments I didn't reply to, I'm fine with.  I'll put together a patch with the changes.

tonyhutter commented:
This is the latest incarnation of the patch.  I think I got everything @behlendorf was looking for:

- Don't assign "uint64_t ishole" to bool
- Remove "goto end" in for_each_vdev_cb() and return directly instead
- Cleaned up "HELP_IOSTAT" string. Added newlines, "[-T d | u]" and "[interval[count]]" sections.
- s/times labels/time labels/ in zpool.8 man page
- Rename zfs_nicenum_impl -> zfs_nicenum_format
- Fix whitespace in zfs_nicenum_impl()
- Added spaces between some variable declerations and body of blocks
- Used verify() directly on nvlist_lookup_uint64_array() in calc_and_alloc_stats()
- Make single_histo_average() function comment multi-line.
- Make type_to_nvlist[] entries one on each line
- Add zfs_niceraw() for completeness
- ffsll()-> lowbit64()
- Use umem_zalloc()
- Don't check for NULL after kmem_alloc(,...KM_SLEEP)
- Remove ZPOOL_CONFIG_VDEV_STATS_FLAGS
- Add "_VDEV_" in stats nvlist names
- Remove ZIO_PRIORITY_COUNT
- Remove extra line
- Remove zio->io_disk_delay
- Fix get_stat_flags() function comment to make it clear the mask returned are the stats that are supported by both the module and zpool iostat.
- Split queues up into individual nvpairs.
- Group extended stats into their own nvlist within the main list
- Remove unneeded spa_update_l2cache_stats_ex() function, and use vdev_config_generate_stats() instead.
- Remove unnecessary zio->io_delay "start-timer" calls

behlendorf commented on lib/libzpool/util.c:
```diff
@@ -76,6 +76,12 @@ show_vdev_stats(const char *desc, const char *ctype, nvlist_t *nv, int indent)
+		printf("%s: not enough memory\n", __func__);
+		return;
+	}
+
```
UMEM_NOFAIL really does mean it can't fail and the error handling can be dropped.  You'll also want to free this with `umem_free()` which does just call `free()`.

behlendorf commented:
Aside from the minor umem_alloc() nit this looks good to me. I like how placing all the extended stats in their own nvlist simplified things.

So the only remaining missing piece is to add a few test cases to `tests/zfs-tests/tests/functional/cli_user/zpool_iostat/`.  I'd suggest adding at least the following very simple tests.

* Verify that all existing and new `zpool iostat` options are accepted by the command and check for a success exit code.

* Verify that the various permutations of `zpool iostat <pool> <vdev> <vdev>` all work and check for a success exit code.

I've already done some manual testing with the patch, but I'll put it on a more interesting system and give it a work out to make sure the reported results are right.  @richardelling @dweeezil if you two have some time it would be great if you could comment/test this too.  You've both done a ton of performance work with ZFS and we'd love to get your feedback.

tonyhutter commented:
Sounds good, I'll fix the umem_alloc() and add the tests.

dweeezil commented:
I gave this a try on a large system and experimented with adding various almost-defective drives to the pool.  So far, I'd have to say this looks very good.  I was easily able to spot the bad drive when using the new "-l" option.

behlendorf commented on include/sys/fs/zfs.h:
```diff
@@ -741,6 +766,11 @@ typedef enum zpool_errata {
+/* Include ARRAY_SIZE for userspace */
+#ifndef ARRAY_SIZE
+#define	ARRAY_SIZE(a) (sizeof (a) / sizeof (a[0]))
+#endif
```
ARRAY_SIZE is a Linux kernel specific `#define`.  How about instead of using it directly we add ZFS_ARRAY_SIZE to `zfs_context.h` and define it as ARRAY_SIZE for the kernel, and then in the user space portion of this header define it as you've done here.

tonyhutter commented:
Here's the latest patch.  It fixes the issues from the comments:

- Add zpool_iostat_004_pos.ksh test
- Move ARRAY_SIZE() to sysmacros.h after further discusssion with @behlendorf
- Remove umem_zalloc() NULL checks

I also talked offline with @behlendorf about some additional nice-to-have features, and added them in:

- Add stats for "pending" queue and revamp "-q" output.  We can now see both the "active" queue (IOs issued to disk) and "pending" queue (IOs queued up waiting to be issued):


```
                         capacity     operations     bandwidth    syncq_read    syncq_write   asyncq_read  asyncq_write   scrubq_read
pool                   alloc   free   read  write   read  write   pend  activ   pend  activ   pend  activ   pend  activ   pend  activ
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool                  201G   171G      0  1.11K      0  29.1M      0      0      0      0      0      0    207     19      0      0
  A1                    100G  85.7G      0    285      0  1.60M      0      0      0      0      0      0     88      9      0      0
  A2                    100G  85.6G      0    473      0  3.21M      0      0      0      0      0      0    118      9      0      0
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A3                   1.40G   185G      0    374      0  24.3M      0      0      0      0      0      0      0      0      0      0
cache                      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A4                    113K   186G      0      0      0      0      0      0      0      0      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool2                 401G   343G      0  4.54K      0   308M      0      0      0     78      0      0      0      0      0      0
  raidz2                401G   343G      0      0      0      0      0      0      0      0      0      0      0      0      0      0
    A5                     -      -      0      0      0      0      0      0      0      0      0      0      0      0      0      0
    A6                     -      -      0      0      0      0      0      0      0      0      0      0      0      0      0      0
    A7                     -      -      0      0      0      0      0      0      0      0      0      0      0      0      0      0
    A8                     -      -      0      0      0      0      0      0      0      0      0      0      0      0      0      0
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A9                   2.02G   184G      0  4.54K      0   308M      0      0      0     78      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool3                 180G  5.81G      0      0      0      0      0      0      0      0      0      0      0      0      0      0
  A10                   180G  5.81G      0      0      0      0      0      0      0      0      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
```


- Allow the "interval" value to be floating point.  This gives you much higher refresh rates, and is really needed to see the change in the queue values. You can now do things like:

`    zpool iostat 0.5`


- Include average latency stats for queues:

```
                         capacity     operations     bandwidth    total_wait     disk_wait    syncq_wait    asyncq_wait  scrub
pool                   alloc   free   read  write   read  write   read  write   read  write   read  write   read  write   wait
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool                  201G   171G     11    114  1.37M  7.18M  336ns    6us  105ns  144ns      -    6us  268ns  339ns   54ns
  A1                    100G  85.7G      5      4   701K  75.9K  338ns  425ns  105ns  111ns      -      -  269ns  347ns   67ns
  A2                    100G  85.6G      5      4   702K  65.6K  335ns  410ns  104ns  113ns      -      -  268ns  331ns   36ns
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A3                    128K   186G      0    105     99  7.04M   14ns    6us   13ns  147ns    1ns    6us      -      -      -
cache                      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A4                    146K   186G      0      0      9  2.25K   11ns  210ns   11ns  115ns      -      -      -   96ns      -
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool2                 401G   343G     15    135  1.37M  7.79M  212ns    6us  107ns  140ns      -    7us  129ns  407ns   39ns
  raidz2                401G   343G     15     23  1.37M   269K  213ns  471ns  108ns   96ns      -      -  129ns  407ns   39ns
    A5                     -      -      3      5   351K  67.2K  213ns  478ns  108ns   96ns      -      -  129ns  414ns   32ns
    A6                     -      -      3      5   351K  67.2K  213ns  469ns  107ns   97ns      -      -  129ns  404ns   32ns
    A7                     -      -      3      5   351K  67.2K  214ns  465ns  108ns   95ns      -      -  129ns  401ns   49ns
    A8                     -      -      3      5   351K  67.2K  211ns  473ns  107ns   96ns      -      -  128ns  408ns   45ns
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A9                    128K   186G      0    112     99  7.53M    4ns    7us    4ns  149ns      -    7us      -      -      -
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool3                 180G  5.81G     11      0  1.37M    134  655ns  168ns  101ns   18ns      -      -  641ns  170ns   49ns
  A10                   180G  5.81G     11      0  1.37M    134  655ns  168ns  101ns   18ns      -      -  641ns  170ns   49ns
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----

```

- Fixed garbled output when pool names are longer than vdev names with -v invoked.  iostat was only looking at vdev names (and not pool names) when calculating column size with "-v". This seems to be an existing bug in the iostat code:
  
```
    # Before fix
    zpool iostat -v
    ...
                   capacity     operations    bandwidth
    pool        alloc   free   read  write   read  write
    ----------  -----  -----  -----  -----  -----  -----
    testpool.2900   120K  1.98G      0      0     10    785
      loop2      120K  1.98G      0      0     10    785
    ----------  -----  -----  -----  -----  -----  -----
    testpool.3703   120K  1.98G      0      0     13  1.00K
      loop3      120K  1.98G      0      0     13  1.00K
    ----------  -----  -----  -----  -----  -----  -----
    testpool.4502   112K  1.98G      0      0     23  1.79K
      loop4      112K  1.98G      0      0     23  1.79K
    ----------  -----  -----  -----  -----  -----  -----
    
    # After fix
    zpool iostat -v
    ...
                     capacity     operations     bandwidth
    pool           alloc   free   read  write   read  write
    -------------  -----  -----  -----  -----  -----  -----
    testpool.2900   120K  1.98G      0      0      9    680
      loop2         120K  1.98G      0      0      9    680
    -------------  -----  -----  -----  -----  -----  -----
    testpool.3703   120K  1.98G      0      0     11    855
      loop3         120K  1.98G      0      0     11    855
    -------------  -----  -----  -----  -----  -----  -----
    testpool.4502   112K  1.98G      0      0     17  1.33K
      loop4         112K  1.98G      0      0     17  1.33K
    -------------  -----  -----  -----  -----  -----  -----
```

richardelling commented:
@tonyhutter this is really looking good. By chance do you have an output during a replace? That has been known to cause more formatting challenges.

tonyhutter commented:
@richardelling, the replace formatting works for me with the new code, but it also worked for me for vanilla iostat, so I dunno.  Let me know if you have a test that will reproduce the bad formatting.

```
                         capacity     operations     bandwidth    total_wait     disk_wait    syncq_wait    asyncq_wait  scrub   syncq_read    syncq_write   asyncq_read  asyncq_write   scrubq_read
pool                   alloc   free   read  write   read  write   read  write   read  write   read  write   read  write   wait   pend  activ   pend  activ   pend  activ   pend  activ   pend  activ
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool                  201G   171G  2.28K  2.27K   288M   287M    4ms   99ms    1ms    1ms      -      -      -   99ms    2ms      0      0      0      0      0      0    178      1      5      3
  A0                    100G  85.7G     17      0  17.9K      0    6ms      -    6ms      -      -      -      -      -    1us      0      0      0      0      0      0      0      0      0      0
  replacing             100G  85.6G  2.26K  2.27K   288M   287M    4ms   99ms    1ms    1ms      -      -      -   99ms    2ms      0      0      0      0      0      0    178      1      3      3
    A1                     -      -  2.26K      0   288M      0    4ms      -    1ms      -      -      -      -      -    2ms      0      0      0      0      0      0      0      0      3      1
    A2                     -      -      0  2.27K      0   287M      -   99ms      -    1ms      -      -      -   99ms      -      0      0      0      0      0      0    180      1      0      0
logs                       -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A3                    128K   186G      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
cache                      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -      -
  A4                   35.5K   186G      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
mypool2                 401G   343G      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
  raidz2                401G   343G      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
    A5                     -      -      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
    A6                     -      -      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
    A7                     -      -      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
    A8                     -      -      0      0      0      0      -      -      -      -      -      -      -      -      -      0      0      0      0      0      0      0      0      0      0
---------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
```

behlendorf commented:
@ahrens @adilger I'd love to get your feedback on this first round of extensions to the `zpool iostat` command.  Does the output intuitively make sense, are the options clearly described in the man page, is it structed in a way that it won't cause issues when submitted upstream.

adilger commented:
Brian, I made a pass through the patch, but can't really comment on the structure of the code since it isn't an area that I've looked at before.  The added information definitely looks useful, especially the per-disk latency.

I recall there was going to be an option to generate standard formatted output (YAML?) to make it easier to parse such data.  I don't think the current "-H" format meets those goals, mostly because it isn't very easy to parse values with attached units, so it should be possible to disable units and just get the raw values (I didn't see where zfs_nice*() was being called so I couldn't see if that is being done).

ahrens commented:
This output looks super useful.  Couple of questions before I dive into the code:
* Can you also do histograms of latency per device?
* Does the parsable (-Hp) output of histogram show the buckets all in ns?  I imagine this would make it clear why the buckets are odd numbers of milliseconds.
* When not doing parsable, it might make sense for the histogram to omit the first and last chunks of all-zero buckets (AFAIK nobody has 1ns latency so this bucket is always empty, it just takes up real estate).

behlendorf commented:
@adilger thanks for looking.  The JSON work your referring to is in #3938 and is designed to further extend the commands with a -j/-J option to output standard JSON to be fed in to other tools.  That works moving along and they're working on adding test cases for the new JSON output.

> Can you also do histograms of latency per device?

Yes, you can specify a specific vdev with the pool for example: `zpool iostat -G dozer sda`.

tonyhutter commented:
Thanks for looking this over guys.  A couple of answers to your questions:

@adilger, the -H option can already be passed to "zpool events" "zpool get" and "zpool list", so I just kept that convention for "zpool iostat".  Note that with iostat, you can you can also pass in -p ("parseable") with -H to get the raw values. I agree that YAML/json output would be very cool too.

@ahrens, as @behlendorf pointed out, you can pass in a specific vdev for stats.  This works for all the stats, not just histograms.  You can even pass in a list of vdevs from different pools.

You're correct that -Hp on histograms prints out nanosecs.  The first column label is the end of the histogram bucket.  So for example when you see the 8us (8191) label, it's saying the bucket covers 4us-8us (4096ns-8191ns):

```
-G
8us             0      4      0     16      5     24      0     44      0

-GH
8191	0	4	0	16	5	24	0	44	0
```
I don't really have a strong opinion on omitting the leading/ending all-zero buckets on the histograms.  It would take up less screen real estate, which is good.  However, it would also cause occasional "jumps" in the output if you were live viewing the histograms with `interval` set (basically when using it as a poor-man's 'top').   I also worry a little about having different buckets shown for the parseable and niceprint output; seems like they should be the same for consistency.


On an unrelated note, I just pushed a small fix into this patch to remove an extra separator line being printed when -H was set but -v was not:

```
$ zpool  iostat -Hp
mypool	111616	2080263168	0	0	8	41
mypool2	167424	486371840	0	0	17	54
----------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
```

behlendorf commented:
> opinion on omitting the leading/ending all-zero buckets on the histograms.

I can see the potential appeal to omitting the leading/ending all-zero buckets.  However, for the reasons @tonyhutter mentioned I think this should be optional behavior.  One possible solution would be to leverage the verbose option so that all-zero buckets are only printed when **-v** is passed.  Although I'm wary of further overloading that option.

Alternately, we use a different letter and use lower and upper case to allow both formats, say **-w** and **-W**.  I suggest we change **-G** to **-W** which makes more logical sense to me since we're looking at I/O wait times.  Since **-w** isn't used we can always come back at some point in the future and use this option to implement a version which trims all-zero buckets if needed.  Or implement it some other way entirely, this leaves our options open.

tonyhutter commented:
@behlendorf I pushed the '-G' -> '-w' rename; I agree that the ''w' makes more sense for latency ("wait").  I went with a lowercase 'w' to make it easier to type.  

In the future we could always add in a '-z' to omit the leading/trailing zero rows.  That would kinda, sorta, follow the convention of the non-ZFS `iostat` command:
```
-z
    Tell iostat to omit output for any devices for which there was no activity during the sample period.
```
Let's leave that for another commit though, since this patch is getting pretty big as-is.

behlendorf commented:
@tonyhutter sounds good.  OK, then this LGTM.  For anyone interested in this change it would be great to get a few more people to sign off on it.  Hopefully we can then get in merged in the next day or two.

behlendorf commented:
Merged as:

193a37c Add -lhHpw options to "zpool iostat" for avg latency, histograms, & queue

lundman commented:
Not a big deal, but it steps on the BSD toes with `isnumber()` which takes an `int` (char) instead of a `char *`. I renamed it `zfs_isnumber()`
Also, I think you guys just made up `ARRAY_SIZE` :)  Not convinced my clang compatible version does the proper typechecking, but I grabbed something off stackoverflow.

behlendorf commented:
News to me about `isnumber()`!  `ARRAY_SIZE()` is a Linux kernel thing.  :)
